### PR TITLE
Fix Issue #829 (Allow Clearance::BackDoor in development environments)

### PR DESCRIPTION
**Pull request by nsuchy**
This pull request merges a fix which fixes Issue #829. The developer can now specify an array of allowed environments to allow Clearance::BackDoor. I modified existing test cases to account for the changes I made (perhaps I should add more tests?).

For further context see: https://github.com/thoughtbot/clearance/issues/829

### CODE REVIEW

File: lib/clearance/back_door.rb [hunk 1]
```diff
@@ -31,8 +31,9 @@ module Clearance
   #   visit new_feedback_path(as: user)
   class BackDoor
     def initialize(app, &block)
-      unless ENV["RAILS_ENV"] == "test"
-        raise "Can't use backdoor outside test environment"
+      @allowed_envs = Clearance.configuration.allowed_envs || ["test", "ci", "development"]
```
[the remark this answers is not in the document]
Comment (mjankowski) [reply lib/clearance/back_door.rb]: We probably don’t need the default set here, it will already come from the configuration initialization.

Comment (mjankowski) [inline lib/clearance/back_door.rb]: If we set these defaults in the config itself, we probably don't need to set here?

File: lib/clearance/back_door.rb [hunk 2]
```diff
@@ -31,8 +31,9 @@ module Clearance
...
+      unless @allowed_envs.include? ENV["RAILS_ENV"]
```
Comment (mjankowski) [inline lib/clearance/back_door.rb]: Can we skip the the i-var and use the config directly? Maybe wrap up the check in a well named private method?

File: lib/clearance/configuration.rb [hunk 3]
```diff
@@ -90,6 +90,11 @@ class Configuration
     # @return [ActiveRecord::Base]
     attr_accessor :user_model
 
+    # The array of explictly allowed environments to use Clearance::BackDoor in.
+    # Defaults to ["test", "ci", "development"]
+    # @return [Array<String>]
+    attr_accessor :allowed_envs
```
Comment (mjankowski) [inline lib/clearance/configuration.rb]: Can we include "backdoor" in this config name so that it's clear what the envs are about?

Comment (nsuchy) [reply lib/clearance/configuration.rb]: See the new name set in my changes.

File: lib/clearance/back_door.rb (hunk 2 above)
Comment (nsuchy) [reply lib/clearance/back_door.rb]: I agree. It now just checks the variable directly and assumes it's working as expected. Defaults were set in the configuration initializer anyways.

File: lib/clearance/back_door.rb (hunk 1 above)
Comment (nsuchy) [reply lib/clearance/back_door.rb]: Agreed, resolved.

**Comment by nsuchy** [comment]
Unless I'm missing something in the issue, this PR will also resolve https://github.com/thoughtbot/clearance/issues/825

File: lib/clearance/back_door.rb [hunk 4]
```diff
@@ -31,8 +31,8 @@ module Clearance
   #   visit new_feedback_path(as: user)
   class BackDoor
     def initialize(app, &block)
-      unless ENV["RAILS_ENV"] == "test"
-        raise "Can't use backdoor outside test environment"
+      unless Clearance.configuration.allowed_backdoor_environments.include? ENV["RAILS_ENV"]
```
Comment (mjankowski) [inline lib/clearance/back_door.rb]: I realize it wasn't like this before you changed - but can we extract this to private method?

**Comment by mjankowski** [comment]
I left a few what are probably final notes -- if you also want to a) rebase/squash this against master, b) add something to the NEWS file about this change - I'll merge in.

**Comment by nsuchy** [comment]
For rebase/squash, I generally work on a branch and prefer to keep version history, unless there's a pressing reason to squash on my end can you use squash and merge?

**Comment by mjankowski** [comment]
No prob - when your changes are good I can massage the commits to my liking.

File: lib/clearance/back_door.rb (hunk 4 above)
Comment (nsuchy) [reply lib/clearance/back_door.rb]: Are you asking to make initialize a private method or the unless check a private method?

Additionally for the comment are you asking me to add extra unit tests to check different envs?

**Comment by nsuchy** [comment]
@mjankowski Awesome, a commit history on a branch often makes it easier when going back to audit a pull request to see what the exact thought process was during development even if it does leave the PR itself a bit messy. I requested two clarifications and will make the changes after you reply 😄

File: lib/clearance/back_door.rb (hunk 4 above)
Comment (mjankowski) [reply lib/clearance/back_door.rb]: Private method within `BackDoor` for the check - something like `unless allowed_backdoor_env?` or similar.

And yes - a unit test that shows the configuration overriding the default set of envs.

Comment (nsuchy) [reply lib/clearance/back_door.rb]: I see I'll make a private method to check the environment and call it instead.

I'm not too familiar with rspec though I came up with this based on the other tests? Would this work for our needs?
```ruby
it "can be used with non-default explicitly allowed envs" do
    Clearance.configuration.allowed_backdoor_environments.push("demo")
    with_environment("RAILS_ENV" => "demo") do
      user_id = "123"
      user = double("user")
      allow(User).to receive(:find).with(user_id).and_return(user)
      env = env_for_user_id(user_id)
      back_door = Clearance::BackDoor.new(mock_app)

      result = back_door.call(env)

      expect(env[:clearance]).to have_received(:sign_in).with(user)
      expect(result).to eq mock_app.call(env)
    end 
  end
```

**Comment by nsuchy** [comment]
Okay so it appears that the unit tests were successful 😄

Is this ready for a merge or are there any final changes you'd like me to make?

**Comment by mjankowski** [comment]
I rebased and did small cleanup, merged in here: https://github.com/thoughtbot/clearance/commit/b77d86622410a753ec3843333c29787845562c1b

While I was in there I:

- Added a case for totally disabling the backdoor by setting the config option to `nil`
- Added news file entry
- Changed error message to raise a message which includes the configured envs (easier debug if misconfigured).

Thanks!

**Pull request closed without merging**